account/recovery.go

```diff
@@ -220,16 +379,121 @@ func recoverInitialState(ctx context.Context, cfg RecoveryConfig) ([]*Account,
 func updateAccountStates(cfg RecoveryConfig, client *rpcclient.Client,
 	accounts []*Account) ([]*Account, error) {
 
-	// TODO (positiveblue): update account states
+	for i := cfg.FirstBlock; i < uint32(cfg.CurrentBlockHeight); i++ {
+		blockHeight := i
+
+		blockHash, err := client.GetBlockHash(int64(i))
+		if err != nil {
+			return nil, err
+		}
+
+		block, err := client.GetBlock(blockHash)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, tx := range block.Transactions {
+			for idx, acc := range accounts {
+				// It's safe to ignore accounts that are in a
+				// terminal state.
+				if acc.State == StateClosed {
+					continue
+				}
+				found := poolscript.IncludesPreviousOutPoint(
+					tx, acc.OutPoint,
+				)
+
+				if !found {
+					continue
+				}
+
+				newAcc, err := findAccountUpdate(
+					cfg, acc, blockHeight, tx,
+				)
+				if err != nil {
+					log.Debugf("unable to find account "+
+						"update: %v", err)
+					continue
+				}
+
+				accounts[idx] = newAcc
+			}
+		}
+	}
+	return accounts, nil
+}
+
+// findAccountUpdate tries to find the new account values after an update.
+func findAccountUpdate(cfg RecoveryConfig, acc *Account, blockHeight uint32,
+	tx *wire.MsgTx) (*Account, error) {
+
+	newAcc := acc.Copy()
+	newAcc.BatchKey = poolscript.IncrementKey(newAcc.BatchKey)
+
+	idx, ok := matchScript(newAcc, newAcc.Expiry, tx)
+	if ok {
+		newAcc.Value = btcutil.Amount(tx.TxOut[idx].Value)
+		// TODO (positiveblue): change 0 for dust limit.
+		if newAcc.Value == 0 {
```

Comment:
We never re-create an account below the dust value. So if an account is fully spent (e.g. its balance falls below the dust threshold), its account isn't re-created. In that case we won't be able to find an output for it on chain. See comment above about just logging and assuming it to be closed.